Sources/SKSwiftPMWorkspace/SwiftPMBuildSystem.swift

```diff
@@ -314,43 +323,72 @@ extension SwiftPMBuildSystem: SKCore.BuildSystem {
 
   public var indexPrefixMappings: [PathPrefixMapping] { return [] }
 
-  public func buildSettings(for uri: DocumentURI, language: Language) throws -> FileBuildSettings? {
-    // SwiftPMBuildSystem doesn't respect the langue specified by the editor.
-    return try buildSettings(for: uri)
-  }
-
-  private func buildSettings(for uri: DocumentURI) throws -> FileBuildSettings? {
-    guard let url = uri.fileURL else {
+  public func buildSettings(
+    for uri: DocumentURI,
+    in configuredTarget: ConfiguredTarget,
+    language: Language
+  ) throws -> FileBuildSettings? {
+    guard let url = uri.fileURL, let path = try? AbsolutePath(validating: url.path) else {
       // We can't determine build settings for non-file URIs.
       return nil
     }
-    guard let path = try? AbsolutePath(validating: url.path) else {
-      return nil
-    }
 
-    if let buildTarget = try buildTarget(for: path) {
-      return FileBuildSettings(
-        compilerArguments: try buildTarget.compileArguments(for: path.asURL),
-        workingDirectory: workspacePath.pathString
-      )
+    if configuredTarget.targetID == "" {
+      return try settings(forPackageManifest: path)
     }
 
-    if path.basename == "Package.swift" {
-      return try settings(forPackageManifest: path)
+    let buildTargets = self.targets.filter({ $0.name == configuredTarget.targetID })
+    if buildTargets.count > 1 {
+      logger.error("Found multiple targets with name \(configuredTarget.targetID). Picking the first one")
+    }
+    guard let buildTarget = buildTargets.first else {
+      if buildTargets.isEmpty {
+        logger.error("Did not find target with name \(configuredTarget.targetID)")
+      }
+      return nil
     }
 
-    if path.extension == "h" {
-      return try settings(forHeader: path)
+    if url.pathExtension == "h", let substituteFile = buildTarget.sources.first {
+      return FileBuildSettings(
+        compilerArguments: try buildTarget.compileArguments(for: substituteFile),
+        workingDirectory: workspacePath.pathString
+      ).patching(newFile: try resolveSymlinks(path).pathString, originalFile: substituteFile.absoluteString)
```

Comment:
Do we actually want to only compile the header? Has pros and cons - I imagine there's a bunch of codebases where the header doesn't actually compile in isolation, but it would generally be much faster.

Reply:
We don’t actually compile the headers. The primary use case of this function is to provide the compiler arguments for clangd for semantic functionality. During indexing, we always index a main file. Not sure if that code existed when you reviewed the PR though.

https://github.com/apple/sourcekit-lsp/blob/490871b21f90bc1270fb4c57978b2ab0b14073e9/Sources/SemanticIndex/SemanticIndexManager.swift#L236-L241